### ISSUE TITLE
switching TV connected to HDMI On via CEC

### ISSUE THREAD

**Issue by lolouk44** [miracle2k/onkyo-eiscp#80]
I see in the list of commands that one can enable or disable CEC, but I can't see the command to turn a TV on.
Yet if I turn my blu-ray or satellite receiver on, the TV turns on automatically.

What command would I need to use to turn the TV on?
Thanks

**Comment by ouija** [comment]
Curious as well if there is an HDMI-CEC "Power On" command? 👍

**Comment by ouija** [comment]
Looked into this a bit further;  Unfortunately doesn't look like the ISCP commands for "TV (via RIHD)" have been implemented into the script.

Any way this can be requested? (please!)  Basically looking to sending a "POWER", "PWRON", "PWROFF" command via "CTV" code....  sure make it sound simple... going to see if I can figure out how to do this myself, but my Python stills are terrible..

**Comment by lolouk44** [comment]
@ouija I've not seen either an ISCP command that would do that, but if you do find one, please post your findings here 👍

**Comment by ouija** [comment]
I was curious about this, and came across another utility for sending iscp commands via Linux (http://simon.aldrich.eu/blog/2013/01/onkyo-network-remote/) which allows you to define the actual commands.

So after building the binary, I found that running the command:
`./onkyo-iscp 192.168.1.100 CTV POWER` and `./onkyo-iscp 192.168.1.100 CTV PWROFF` had no effect.

However, running the commands `./onkyo-iscp 192.168.1.100 CTV MUTE` and `./onkyo-iscp 192.168.1.100 CTV VLUP` or `./onkyo-iscp 192.168.1.100 CTV VLDN` did...

**Comment by ouija** [comment]
**UPDATE:** Sorry for turning this thread into a novel, but did some playing around with the source code tonight and successfully have onkyo-eiscp sending the CEC commands, but also think I gained a better understanding of _why_ this was left out of the script to begin with.  

Seems the CEC commands don't return a response from the receiver the way all other commands do (where it matches the response to the group of commands it was being called from).  

For example, sending a volume up command via CEC returns a response similar to a **master-volume=level-up** command, but this response doesn't match the same command group as the initial CEC command, and the script will end up timing out.

The CEC command is still sent and it works, but the error message isn't pretty and leaves the impression that it fails.

Sooo... I decided to try and implement support for the CEC commands myself tonight _(since @miracle2k  wrote this thing so brilliantly that it's actually really easy to add in additional commands)_

The bigger issue was with the script timing out due to the responses not existing under the group of the cec commands (and really a lack of a response at all to CEC commands, other than the initial command being sent),

But once I figured out how the responses were being handled, I added an exception to the script which essentially just returns what CEC commands were issued (instead of waiting for a response to these commands) and it now works beautifully!!

**I now am able to turn on my TV and Receiver using a single command: `onkyo power=on hdmi-cec.ctv=on`**

You can find my fork with all the changes at: https://github.com/ouija/onkyo-eiscp

If you simply replace the `commands.py` and `core.py` files under the eiscp package directory for python on your machine, you should be good to go.

Thanks to @miracle2k for making such a useful script, and also thanks to Simon Aldrich for his script as well!

**Comment by lolouk44** [comment]
Great news, I can't wait to try tonight :)

**Comment by miracle2k** [comment]
@ouija Looks like something I might be able to merge; I'll try to find the time to look at it. 

Note that with this script you can also submit raw commands like PWR01.

**Comment by ouija** [comment]
Awesome.   Yeah, after I became more familiar with the script I did realize this.  Great job on it, it's going to serve my needs very well!

**Comment by lolouk44** [comment]
Hi @miracle2k I was wondering if you could share with us approx when you think you could merge @ouija 's code?
I'd rather wait for your version than play with different repos.
Also I take it you're the owner of the [Home-Assistant](https://github.com/home-assistant/home-assistant) eiscp code? If so will you also update it?
Thanks

**Comment by wrobelda** [comment]
Just tested this with my BenQ projector and while turning off works fine, turning off doesn't. I guess it's due to CEC manufacturer incompatibilities.

**Comment by ouija** [comment]
Note there are two possible CEC power commands that can be issued: `onkyo hdmi-cec.ctv=on` and `onkyo hdmi-cec.ctv=power`

The specification had two separates codes listed for TV CEC Power _(PWRON and POWER)_ and each of these commands send either code, which may make a difference in your case, but not sure.

 I think the single POWER command can do both power on and off (haven't really tested) and the ISCP documentation doesn't provide any details other can the code names :)

**Comment by wrobelda** [comment]
> but note there are three possible CEC power commands that can be issued: onkyo hdmi-cec.ctv=on, onkyo hdmi-cec.ctv=off and onkyo hdmi-cec.ctv=power

Yeah, I have in fact tried them all, but still, only the "on" command works.

> Also try maybe the CEC DVD Commands? ie: onkyo hdmi-cec.cdv=on
That's what I use now. "on" switches on my Apple TV, receiver and projector. "off" switches the Apple TV and a receiver, but not a projector. Similarly, if I turn off my Apple TV manually using its remote, only the receiver will turn off. I guess this is, therefore, BenQ's incompatible implementation here causing this behavior.

I worked around, however, but setting the auto power-off on the projector to 5 mins when no signal is present and that basically does the job.

**Comment by toshibochan** [comment]
Did you try change off to standby?

**Comment by wrobelda** [comment]
> I have in fact tried them all, but still, only the "on" command works.

Yeah.

**Comment by rui-nar** [comment]
hey @miracle2k , do you think you are able to merge @ouija commits to enable CEC TV control ?
I'd rather continue using your (official) package (from PyPi) rather than modifying by hand using @ouija modifications.

many thanks

**Comment by toshibochan** [comment]
Did you try onkyo hdmi-cec.ctv=standby

**Comment by misomosi** [comment]
I can't seem to get any of these CEC commands to do anything other than the ones that control volume. None of the power commands perform regardless of the TV state, but any connected devices like a PS4 has no problem turning the TV on.

TV is an LG 55UH7700, so perhaps there are incompatibilities with LG?

**Comment by ouija** [comment]
I had similiar issues with an LG 70UJ6570 and ended up using another script to power on the TV, and created my own fork of this here: https://github.com/ouija/LGWebOSRemote

The TV will send a CEC command to the Onkyo Reciever and power it on as well, if connected to HDMI ARC / CEC Enabled.

Hope this helps!

**Comment by misomosi** [comment]
> 
> 
> I had similiar issues with an LG 70UJ6570 and ended up using another script to power on the TV, and created my own fork of this here: https://github.com/ouija/LGWebOSRemote
> 
> The TV will send a CEC command to the Onkyo Reciever and power it on as well, if connected to HDMI ARC / CEC Enabled.
> 
> Hope this helps!

Excellent workaround, ouija! I wasn't aware you could power the TV on and off from the network.
